src/resolve/resolved-dnssd.c

```diff
@@ -253,6 +262,16 @@ int dnssd_update_rrs(DnssdService *s) {
         if (!s->ptr_rr->ptr.name)
                 goto oom;
 
+        s->sub_ptr_rr = dns_resource_record_new_full(DNS_CLASS_IN, DNS_TYPE_PTR,
+                                                 selective_name);
```

Comment:
Looks like services without subtypes crash resolved:
```sh
Using system hostname 'H'.                                                                                                                                    
Assertion 'name' failed at src/resolve/resolved-dns-rr.c:24, function dns_resource_key_new(). Aborting.
...
#0  0x00007ffff56b0884 in __pthread_kill_implementation () from /lib64/libc.so.6
#1  0x00007ffff565fafe in raise () from /lib64/libc.so.6
#2  0x00007ffff564887f in abort () from /lib64/libc.so.6
#3  0x00007ffff71074f1 in log_assert_failed (text=0x8a8720 <str> "name", file=0x8ac684 "src/resolve/resolved-dns-rr.c", line=24, 
    func=0x8a8780 <__func__.dns_resource_key_new> "dns_resource_key_new") at ../src/basic/log.c:968
#4  0x00000000007cc62d in dns_resource_key_new (class=1, type=12, name=0x0) at ../src/resolve/resolved-dns-rr.c:24
#5  0x00000000007d2feb in dns_resource_record_new_full (class=1, type=12, name=0x0) at ../src/resolve/resolved-dns-rr.c:393
#6  0x0000000000692e89 in dnssd_update_rrs (s=0x60800000cba0) at ../src/resolve/resolved-dnssd.c:265
#7  0x00000000006909dd in dnssd_service_load (manager=0x619000000f80, filename=0x6030001541b0 "/etc/systemd/dnssd/qotd.dnssd")
    at ../src/resolve/resolved-dnssd.c:131
#8  0x000000000068d667 in dnssd_load (manager=0x619000000f80) at ../src/resolve/resolved-dnssd.c:203
#9  0x00000000006f01c6 in manager_new (ret=0x7ffff2609020) at ../src/resolve/resolved-manager.c:629
#10 0x000000000076c056 in run (argc=1, argv=0x7fffffffe4b8) at ../src/resolve/resolved.c:72
#11 0x000000000076ab0e in main (argc=1, argv=0x7fffffffe4b8) at ../src/resolve/resolved.c:99
```
Services with subtypes seem to trigger a memory leak:
```sh
==140469==ERROR: LeakSanitizer: detected memory leaks
Direct leak of 112 byte(s) in 1 object(s) allocated from:
    #0 0x4c2f12 in malloc (/home/vagrant/systemd-asan/build/systemd-resolved+0x4c2f12) (BuildId: 72669fc0ab2b83b1baac44fc30ddde9a38aecd33)
    #1 0x7cebc1 in malloc_multiply /home/vagrant/systemd-asan/build/../src/basic/alloc-util.h:119:16
    #2 0x7d2c61 in dns_resource_record_new /home/vagrant/systemd-asan/build/../src/resolve/resolved-dns-rr.c:375:14
    #3 0x7d3062 in dns_resource_record_new_full /home/vagrant/systemd-asan/build/../src/resolve/resolved-dns-rr.c:397:16
    #4 0x692e88 in dnssd_update_rrs /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:265:25
    #5 0x6909dc in dnssd_service_load /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:131:13
    #6 0x68d666 in dnssd_load /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:203:21
    #7 0x6f01c5 in manager_new /home/vagrant/systemd-asan/build/../src/resolve/resolved-manager.c:629:13
    #8 0x76c055 in run /home/vagrant/systemd-asan/build/../src/resolve/resolved.c:72:13
    #9 0x76ab0d in main /home/vagrant/systemd-asan/build/../src/resolve/resolved.c:99:1
    #10 0x7f3856e49b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #11 0x7f3856e49c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #12 0x4282c4 in _start (/home/vagrant/systemd-asan/build/systemd-resolved+0x4282c4) (BuildId: 72669fc0ab2b83b1baac44fc30ddde9a38aecd33)
Indirect leak of 47 byte(s) in 1 object(s) allocated from:
    #0 0x4c304c in __interceptor_calloc (/home/vagrant/systemd-asan/build/systemd-resolved+0x4c304c) (BuildId: 72669fc0ab2b83b1baac44fc30ddde9a38aecd33)
    #1 0x7cc6ba in dns_resource_key_new /home/vagrant/systemd-asan/build/../src/resolve/resolved-dns-rr.c:27:13
    #2 0x7d2fea in dns_resource_record_new_full /home/vagrant/systemd-asan/build/../src/resolve/resolved-dns-rr.c:393:15
    #3 0x692e88 in dnssd_update_rrs /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:265:25
    #4 0x6909dc in dnssd_service_load /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:131:13
    #5 0x68d666 in dnssd_load /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:203:21
    #6 0x6f01c5 in manager_new /home/vagrant/systemd-asan/build/../src/resolve/resolved-manager.c:629:13
    #7 0x76c055 in run /home/vagrant/systemd-asan/build/../src/resolve/resolved.c:72:13
    #8 0x76ab0d in main /home/vagrant/systemd-asan/build/../src/resolve/resolved.c:99:1
    #9 0x7f3856e49b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #10 0x7f3856e49c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #11 0x4282c4 in _start (/home/vagrant/systemd-asan/build/systemd-resolved+0x4282c4) (BuildId: 72669fc0ab2b83b1baac44fc30ddde9a38aecd33)
Indirect leak of 19 byte(s) in 1 object(s) allocated from:
    #0 0x4accf1 in strdup (/home/vagrant/systemd-asan/build/systemd-resolved+0x4accf1) (BuildId: 72669fc0ab2b83b1baac44fc30ddde9a38aecd33)
    #1 0x6931e4 in dnssd_update_rrs /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:271:35
    #2 0x6909dc in dnssd_service_load /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:131:13
    #3 0x68d666 in dnssd_load /home/vagrant/systemd-asan/build/../src/resolve/resolved-dnssd.c:203:21
    #4 0x6f01c5 in manager_new /home/vagrant/systemd-asan/build/../src/resolve/resolved-manager.c:629:13
    #5 0x76c055 in run /home/vagrant/systemd-asan/build/../src/resolve/resolved.c:72:13
    #6 0x76ab0d in main /home/vagrant/systemd-asan/build/../src/resolve/resolved.c:99:1
    #7 0x7f3856e49b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #8 0x7f3856e49c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #9 0x4282c4 in _start (/home/vagrant/systemd-asan/build/systemd-resolved+0x4282c4) (BuildId: 72669fc0ab2b83b1baac44fc30ddde9a38aecd33)
SUMMARY: AddressSanitizer: 178 byte(s) leaked in 3 allocation(s).
```